src/ParallelAlgorithms/LinearSolver/AsynchronousSolvers/ElementActions.hpp

```diff
@@ -196,11 +204,21 @@ struct PrepareSolve {
     contribute_to_residual_observation<FieldsTag, OptionsGroup,
                                        ParallelComponent>(box, cache,
                                                           array_index);
-    return {std::move(box)};
+    // Skip steps entirely if the solve has already converged
+    constexpr size_t step_end_index =
+        tmpl::index_of<ActionList, CompleteStep<FieldsTag, OptionsGroup,
```

Comment:
That's reasonable, but here again I'm concerned about maintainability rather than user-facing interface. For instance, if you or someone else needs to insert another action after the one that is currently used as the final action in a step (or before the one used as the start), all of these indices would individually need to be updated. I'd suggest labels, but I won't insist if you still think it doesn't make an improvement.

Reply:
I see your concern and will keep this in mind. Might add labels in another PR.